Doc/#1209 - #1250 - Add documentation page for each application and page template (#1263)
Avaiga/taipy-doc: docs/templates/applications/sdm_app.md

Reviewer: This is a technical documentation. The content is good. However, we are missing functional documentation: 
- exposing our motivation and the benefits of using the template
- explaining the objective of the result application.
- showing and describing the expected result (each page)
- describing the create operation. This part is already done.
- describing how to run the application. This part is already done).
- explaining what the developer should focus on. Creating the config in Python, updating the placeholders, etc.

> Building a Taipy application with multiple complex scenarios usually requires a significant amount of time and effort. This can delay the time-to-market and increase development costs.

Reviewer: ```suggestion
Building a Taipy application with multiple complex scenarios usually requires a
significant amount of time and effort, in particular at the beginning of a project
when the user interface requirements are not completely stable. This can delay the
time-to-market and increase development costs.
```

> To reduce the developer hours and resources needed, Taipy provides a scenario management application template, which leverages Taipy visual elements to speed up bootstrapping a standard application. This template offers several key benefits:

Reviewer: ```suggestion
To reduce the development time, Taipy provides a scenario management application template,
which leverages Taipy visual elements to speed up bootstrapping a standard application.
This template offers several key benefits:
```

- **Accelerated Development**: By leveraging scenario and data management visual elements, developers can quickly bootstrap a standard application, saving significant development time to focus on delivering business value more efficiently.
- **Comprehensive Features**: The template supports a wide range of functionalities, including data visualization, data analysis, simulation, what-if analysis, and job orchestration monitoring, providing a comprehensive solution for various use cases.
- **Customization**: The template is highly customizable with pre-built placeholders for various components, allowing developers to tailor the application to meet specific requirements and use cases.

> - **Deployment-Ready**: With support for Docker deployment and integration with Taipy Enterprise Edition, the template is suitable for both development and production environments.

Reviewer: ```suggestion
- **Accelerated Development**: By leveraging scenario and data management visual elements,
    developers can quickly bootstrap a standard application, saving significant
    development time to focus on delivering business value more efficiently.
- **Comprehensive Features**: The template supports a wide range of functionalities,
    including data visualization and analysis, data integration, simulation, what-if
    analysis, and job orchestration monitoring, providing a comprehensive solution for
    various use cases.
- **Customization**: The resulting application is highly customizable with pre-built
    placeholders for various components, allowing developers to tailor the application
    to meet specific requirements and use cases.
- **Deployment-Ready**: With support for Docker deployment and integration, the template
    provides deployment helpers on both test and production environments.
```

    [Contact us](https://taipy.io/book-a-call){: .tp-btn .tp-btn--accent target='blank' }

## 1. Application root folder

- Specifies the root folder of the application.
- The default value is "taipy_application".

## 2. Application main Python file

- Sets the name of the main Python file (entry point) of the application.
- The default value is "main.py".

## 3. Application title

- Specifies the title displayed in the web application.
- The default value is "Taipy Application".

## 4. With TOML Config

- Specifies whether the application will use TOML configuration files, otherwise the configuration
  will be in a Python file.
- The default value is "No".

## 5. With a new Git repository

- Specifies whether the application will be initialized as a new Git repository.
- The default value is "No".

## 6. With Docker deployment

- Specifies Docker support for the application.
- Options:
    - "No": No Docker support.
    - "For development": Add a minimal version of `Dockerfile` and `docker-compose.yml` for development.
    - "For production": Add a production-ready `Dockerfile` and `docker-compose.yml`.
- The default value is "No".

> # Application description

Reviewer: In order to avoid confusion with the application description, what about:

```suggestion
# Generated project description
```

Author: I don't like the word "project" here.

The generated project or application is an application.
We also refer to the generated application as "application" consistently throughout the doc.

Reviewer: I agree. 

However, I would also like to have another section describing the application not from a generated code standpoint but from an end-user perspective, like the functionalities available from the user interface. This new section should be at the top of the page, and I would like to differentiate the two section names. That's my motivation.

What do you think?

The generated application has the following folder structure:

```plaintext
taipy_application/
├──── algos/
│   ├──── __init__.py
│   └──── algos.py
│
├──── config/
│   ├──── __init__.py
│   └──── config.py
│
├──── pages/
│   ├──── __init__.py
│   ├──── job_page/
│   ├──── scenario_page/
│   └──── root.py
│
├──── .taipyignore
├──── docker-compose.yml
├──── Dockerfile
├──── main.py
└──── requirements.txt
```

> Your application's folder structure may vary depending on the options you selected during the creation process. Here is a brief overview of the key components:

Reviewer: ```suggestion
Your application's folder structure may vary depending on the options you selected during the
creation process. Here is a brief overview of the key components:
```

> - *algos/*: Contains the *algos.py* file, which is designed to contain various Python functions used to configure tasks. You can add your tasks' functions here, and they will be imported into the main application file.
> - *config/*: Contains the *config.py* file, where you can put your application's configuration. The configuration will be imported into the main application file.
> - *pages/*: Contains the pages of the application.
>     - *root.py* is the root page of the application, which layouts the application, including the navigation bar, and a sidebar that has [scenario_selector](../../refmans/gui/viselements/corelements/scenario_selector.md) and [data_node_selector](../../refmans/gui/viselements/corelements/data_node_selector.md) visual elements.
>     - *job_page/* contains the page for job orchestration management via the [job_selector](../../refmans/gui/viselements/corelements/job_selector.md) visual element.
>     - *scenario_page/* contains the page for scenario management that displays the information stored in the selected scenario and the DAG via [scenario](../../refmans/gui/viselements/corelements/scenario.md) and  [scenario_dag](../../refmans/gui/viselements/corelements/scenario_dag.md) visual elements, as well as displays the information for the selected data node via the [data_node](../../refmans/gui/viselements/corelements/data_node.md) visual element.
> - *.taipyignore*: Specifies files to be protected when running the web server. Please refer to the [Taipy documentationProtect private files](../../userman/run-deploy/run/protect_files.md) for more information.
> - *docker-compose.yml* and *Dockerfile*: The Docker configuration for building the application Docker image.
> - *main.py*: The main Python file of the application.
> - *requirements.txt*: Contains the Python dependencies required by the application.

Reviewer: ```suggestion
- *algos/*: Contains the *algos.py* file, designed to contain various Python functions
    used to configure tasks. You can add your tasks' functions here, and they will be
    imported into the main application file.
- *config/*: Contains the *config.py* file, where you can put your application's
    configuration. The configuration will be imported into the main application file.
- *pages/*: Contains the application pages.
    - *root.py* is the root page of the application, which layouts the application.
        It includes a navigation bar, and a sidebar with a
        [scenario_selector](../../refmans/gui/viselements/corelements/scenario_selector.md)
        and a 
        [data_node_selector](../../refmans/gui/viselements/corelements/data_node_selector.md)
    - *job_page/* contains the page for job monitoring using a
        [job_selector](../../refmans/gui/viselements/corelements/job_selector.md).
    - *scenario_page/* contains the page for scenario analysis and data management. It
        shows a scenario, its DAG, and its data nodes. It uses a
        [scenario](../../refmans/gui/viselements/corelements/scenario.md), a
        [scenario_dag](../../refmans/gui/viselements/corelements/scenario_dag.md)
        and a [data_node](../../refmans/gui/viselements/corelements/data_node.md)
        visual elements.
- *.taipyignore*: Specifies files to be protected when running the web server. Please
    refer to the
    [Protect private files](../../userman/run-deploy/run/protect_files.md) page for
    more information.
- *docker-compose.yml* and *Dockerfile*: The Docker configuration for building and 
    running the application as a Docker container.
- *main.py*: The main Python file of the application.
- *requirements.txt*: Contains the Python dependencies required by the application.
```

# Customize the application

Everything in the generated application can be updated to fit your specific application. This includes the code, configuration files, and any other resources. Specifically, in the **algos/** and **config/** folders, there are placeholders that you can customize precisely to your use case.

> 

Reviewer: ```suggestion
Everything in the generated application can be updated to fit your needs. It includes
the Python code, the configuration files, and any other resources. Specifically, in
the **algos/** and **config/** folders, there are placeholders that you can customize
precisely to your use case.

```

<!-- TODO: Explain that some parts are strongly use case dependent and identified as placeholders and are probably a good piece of code to focus on.   -->

## Customizing the tasks' functions

> The *algos/* folder contains the *algos.py* file, which is designed to contain various Python functions used to configure tasks. The functions are strongly use-case dependent, for example, cleaning data, performing analysis, or running simulations, among other tasks.
> 
> In the *algos.py* file, the `clean_data()` method is provided as a placeholder. You can edit/replace the placeholder method `clean_data()` and add your tasks' functions here.

Reviewer: ```suggestion
The *algos/* folder contains the *algos.py* file, designed to contain various Python
functions used to configure tasks. The functions are strongly use-case dependent,
for example, cleaning data, performing analysis, or running simulations, among other
tasks.

In the *algos.py* file, the `clean_data()` method is provided as a placeholder. You can
edit/replace the placeholder method `clean_data()` and add your tasks' functions here.
```

```python title="algos/algos.py"
def clean_data(df, replacement_type):
    df = df.fillna(replacement_type)
    return df

def analyze_data(df):
    ...
    return analysis_results
```

> To configure tasks using these functions, make sure to update the import statements in the *algos/\_\_init\_\_.py* file.

Reviewer: ```suggestion
To configure tasks using these functions, make sure to update the import statements in
the *algos/\_\_init\_\_.py* file.
```

```python title="algos/__init__.py"
from .algos import clean_data, analyze_data
```

> Then when [customizing the configuration](#customizing-the-configuration), you can import these functions and use them to configure various tasks for your application.

Reviewer: ```suggestion
Then when [customizing the configuration](#customizing-the-configuration), you can
import these functions and use them to configure various tasks for your application.
```

> !!! note "Updating the folder and file names"

Reviewer: ```suggestion
??? note "Renaming generated folders or files"
```

>     If you decide to rename the *algos.py* file or the *algos/* folder, make sure to update the import statements in the *algos/\_\_init\_\_.py* file and any other import statements.

Reviewer: ```suggestion
    If you decide to rename the *algos.py* file or the *algos/* folder, make sure to
    update the import statements in the *algos/\_\_init\_\_.py* file and any other
    import statements.
```

## Customizing the configuration

> The *config/config.py* file contains the `configure()` method, which will be called in the main application file to configure the application.

Reviewer: ```suggestion
The *config/config.py* file contains the `configure()` method, which will be called in
the main application file to configure the application.
```

> If your answer to the ["With TOML Config"](#4-with-toml-config) question was "No" (as default), the placeholder configuration will be stored in the `configure()` method [as Python code](../../userman/advanced_features/configuration/advanced-config.md#python-code-configuration). Import the added functions from the *algos/algos.py* file and use them to configure the tasks.

Reviewer: ```suggestion
If your answer to the ["With TOML Config"](#4-with-toml-config) question was "No"
(as default), the placeholder configuration will be stored in the `configure()` method
[as Python code](../../userman/advanced_features/configuration/advanced-config.md#python-code-configuration). 
Import the added functions from the *algos/algos.py* file and use them to configure the tasks.
```

```python title="config/config.py"

from algos import clean_data, analyze_data

def configure():
    ...
```

> If your answer to the ["With TOML Config"](#4-with-toml-config) question was "Yes", the placeholder configuration will be stored in `config.toml` file. Update the `config.toml` file to configure the application to your specific use case, or you can use [Taipy Studio](../../userman/ecosystem/studio/index.md) for generating a TOML file that can be loaded in the `configure()` method.

Reviewer: ```suggestion
If your answer to the ["With TOML Config"](#4-with-toml-config) question was "Yes",
the placeholder configuration will be stored in `config.toml` file. Update the
`config.toml` file to configure the application to your specific use case, or you can
use [Taipy Studio](../../userman/ecosystem/studio/index.md) for generating a TOML
file that can be loaded in the `configure()` method.
```

> !!! note "Loading the TOML configuration"
> 
>     If you have a different name for the TOML file, make sure to update the file name in the `configure()` method.

Reviewer: ```suggestion
!!! note "Loading the TOML configuration"

    If you have a different name for the TOML file, make sure to update the file name in the
    `configure()` method.
```

>  Customizing the pages

Reviewer: ```suggestion
 ## Customizing the pages
```

> The *pages/* folder contains the pages of the application. You can customize the content of the pages to fit your specific application requirements, as well as customize the layout and visual elements.

Reviewer: ```suggestion
The *pages/* folder contains the application pages. You can customize the
content of the pages to fit your specific requirements, as well as customize
the layout and visual elements.
```

> *pages/root.py* defines the layout the application, including the navigation bar, and a sidebar. Here, you can customize the [layout](../../refmans//gui/viselements/generic/layout.md), [scenario_selector](../../refmans/gui/viselements/corelements/scenario_selector.md), and [data_node_selector](../../refmans/gui/viselements/corelements/data_node_selector.md) visual elements to fit your application's design.
> 
> In *pages/job_page/job_page.py*, you can customize the [job_selector](../../refmans/gui/viselements/corelements/job_selector.md) visual element, which lists all jobs of the application and allows users to select and manage them.
> 
> In *pages/scenario_page/scenario_page.py*, you can customize the [scenario](../../refmans/gui/viselements/corelements/scenario.md), [scenario_dag](../../refmans/gui/viselements/corelements/scenario_dag.md), and [data_node](../../refmans/gui/viselements/corelements/data_node.md) visual elements to display the information of the selected scenario and data node, as well as modifying the `notify_on_submission()` method to handle the notification of a scenario's submission.

Reviewer: ```suggestion
*pages/root.py* defines the layout of the application, including the navigation bar,
and a sidebar. Here, you can customize the
[layout](../../refmans//gui/viselements/generic/layout.md),
the [scenario_selector](../../refmans/gui/viselements/corelements/scenario_selector.md), and
the [data_node_selector](../../refmans/gui/viselements/corelements/data_node_selector.md).

In *pages/job_page/job_page.py*, you can customize the
[job_selector](../../refmans/gui/viselements/corelements/job_selector.md) 
visual element, which lists all jobs of the application and allows users to select
and manage them.

In *pages/scenario_page/scenario_page.py*, you can customize the
[scenario](../../refmans/gui/viselements/corelements/scenario.md),
the [scenario_dag](../../refmans/gui/viselements/corelements/scenario_dag.md),
and the [data_node](../../refmans/gui/viselements/corelements/data_node.md) visual
elements to display the information of the selected scenario and data node, as well
as modifying the `notify_on_submission()` method to handle the notification of a
scenario's submission.
```

time-to-market and increase development costs.

To reduce the development time, Taipy provides a scenario management application template,
which leverages Taipy visual elements to speed up bootstrapping a standard application.
This template offers several key benefits:

- **Accelerated Development**: By leveraging scenario and data management visual elements,
    developers can quickly bootstrap a standard application, saving significant
    development time to focus on delivering business value more efficiently.
- **Comprehensive Features**: The template supports a wide range of functionalities,
    including data visualization and analysis, data integration, simulation, what-if
    analysis, and job orchestration monitoring, providing a comprehensive solution for
    various use cases.
- **Customization**: The resulting application is highly customizable with pre-built
    placeholders for various components, allowing developers to tailor the application
    to meet specific requirements and use cases.
- **Deployment-Ready**: With support for Docker deployment and integration, the template
    provides deployment helpers on both test and production environments.

> 

Reviewer: We can add a short section here describing the end-user functionalities available in UI with screenshots.

1. "Application root folder":
    - Specifies the root folder of the application.
    - The default value is "new_application".
2. "Application main Python file"
    - Sets the name of the main Python file (entry point) of the application.
    - The default value is "main.py".
3. "Application title":
    - Specifies the title displayed in the web application.
    - The default value is "Taipy Application".
4. "With TOML Config":
    - Specifies whether the application will use TOML configuration files, otherwise the configuration
        will be in a Python file.
    - The default value is "No".
5. "With a new Git repository":
    - Specifies whether the application will be initialized as a new Git repository.
    - The default value is "No".
6. "With Docker deployment"
    - Specifies Docker support for the application.
    - Options:
        - "No": No Docker support.
        - "For development": Add a minimal version of `Dockerfile` and `docker-compose.yml` for development.
        - "For production": Add a production-ready `Dockerfile` and `docker-compose.yml`.
    - The default value is "No".

## Run the application

> To run the application, change to the newly created folder and run the application using `taipy run main.py`.

Reviewer: ```suggestion
To run the application, change to the newly created folder and run the application using
`taipy run main.py`.
```

> Taipy's scenario management application template lets you scaffold a custom Taipy
> application using scenario management capabilities. It uses
> [a scenario selector](./../../../refmans/gui/viselements/corelements/scenario_selector.md)
> to allow end users to create, manage, and run scenarios directly from the GUI page.

Reviewer: ```suggestion
Taipy's scenario management application template leverages the scenario and data
management visual elements to speed up bootstrapping a standard application. The
resulting application exposes a user interface for scenario creation and management,
data visualization, data analysis, simulation, what-if analysis, and/or job
orchestration monitoring.
```

> 

Reviewer: ```suggestion
# Application description
TODO: Find a better section name. Proposals:

  - "What to do with the app?"
  - "Application description"
  - "Understand the application"
  - "Features"
  - "Application capabilities"

TODO: Quickly describe each page and each visual element.
TODO: Add links to each visual element documentation page.
TODO: Should we build a config example so we can easily take some screenshots?
```

Author: > Should we build a config example so we can easily take some screenshots?

We can describe an example of the config in the `Customize the application` section.

However as discussed, I would not put screenshots here since we already have screenshots in the visual element pages (which are referenced), and it's sensitive to changes

> # Create a Taipy application from the scenario management template

Reviewer: ```suggestion
# How to create the application
```

> To create an application from the Taipy scenario management application template, run
> `taipy create --application sdm` from the CLI, then answer a few questions to customize your application.

Reviewer: ```suggestion
To create the application from the template, change to the folder in which you want to
create the application and run the command `taipy create --application sdm`. Then answer
the few questions to customize your application.
```

```console
$ taipy create --application sdm
[1/6] Application root folder [taipy_application]:
[2/6] Application main Python file [main.py]:
[3/6] Application title [Taipy Application]:
[4/6] With TOML Config? (No):
[5/6] With a new Git repository? (No):
[6/6] Select With Docker deployment
    1 - No
    2 - For development
    3 - For production
    Choose from [1/2/3] (1):
New Taipy application has been created at ./taipy_application

To start the application, change directory to the newly created folder:
    cd ./taipy_application

You can then run the application as follows:
    taipy run main.py
```

> !!! info

Reviewer: ```suggestion
??? info "Default answers"
```

>     In the CLI, the default value for each question is displayed in the square brackets.

Reviewer: ```suggestion
    In the CLI, the default answer for each question is displayed in the square brackets.
```

    [Contact us](https://taipy.io/book-a-call){: .tp-btn .tp-btn--accent target='blank' }

## 1. Application root folder

- Specifies the root folder of the application.
- The default value is "new_application".

## 2. Application main Python file

- Sets the name of the main Python file (entry point) of the application.
- The default value is "main.py".

## 3. Application title

- Specifies the title displayed in the web application.
- The default value is "Taipy Application".

## 4. With TOML Config

- Specifies whether the application will use TOML configuration files, otherwise the configuration
  will be in a Python file.
- The default value is "No".

## 5. With a new Git repository

- Specifies whether the application will be initialized as a new Git repository.
- The default value is "No".

## 6. With Docker deployment

- Specifies Docker support for the application.
- Options:
    - "No": No Docker support.
    - "For development": Add a minimal version of `Dockerfile` and `docker-compose.yml` for development.
    - "For production": Add a production-ready `Dockerfile` and `docker-compose.yml`.
- The default value is "No".

> # Run the application

Reviewer: ```suggestion
# How to run the application
```

To run the application, change to the newly created folder and run the application using `taipy run main.py`.

```console
$ cd ./taipy_application
$ taipy run main.py
```

If the newly created application supports Docker, you can also run the application using `docker-compose`.

```console
$ cd ./taipy_application
$ docker-compose up --build -d
```

> You can now access the application in your browser at `http://localhost:5000`.

Reviewer: ```suggestion
You can now access the application in your browser at `http://localhost:5000`.

# Customize the application.

TODO: Explain that everything can be updated.
TODO: Explain that some parts are strongly use case dependent and identified as placeholders and are probably a good piece of code to focus on.
TODO: List and explain the various placeholders 

```
